source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+import re
+from pathlib import Path
+from typing import \
+    List, \
+    Mapping, \
+    Tuple, \
+    Type
+from fab.artifact import \
+    Artifact, \
+    FileType, \
+    State, \
+    New, \
+    Unknown, \
+    Analysed, \
+    Compiled, \
+    Linked
+from fab.tasks import Task
+from fab.database import \
+    SqliteStateDatabase, \
+    FileInfoDatabase
+
+
+class PathMap(object):
+    def __init__(self,
+                 pattern: str,
+                 filetype: Type[FileType],
+                 state: Type[State]):
+        self._pattern = pattern
+        self._filetype = filetype
+        self._state = state
+
+    @property
+    def pattern(self):
+        return self._pattern
+
+    @property
+    def filetype(self):
+        return self._filetype
+
+    @property
+    def state(self):
+        return self._state
+
+    def match(self, path: Path) -> bool:
+        matched = False
+        if re.match(self.pattern, str(path)):
```

Comment:
We should compile this pattern in the constructor for faster matching. Please raise a ticket for this.

Reply:
I didn't want to raise an issue/ticket for something this minor so I've done it